bentoml/_internal/exportable.py

```diff
@@ -0,0 +1,363 @@
+import os
+import re
+import typing as t
+import urllib.parse
+from abc import ABC
+from abc import abstractmethod
+import logging
+
+import fs
+import fs.copy
+import fs.mirror
+import fs.opener
+from fs.base import FS
+
+T = t.TypeVar("T")
+
+
+uriSchemeRe = re.compile(r".*[^\\](?=://)")
+
+
+class Exportable(ABC):
+    @staticmethod
+    @abstractmethod
+    def _export_ext() -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def _export_name(self) -> str:
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def _fs(self) -> FS:
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def from_fs(cls: t.Type[T], item_fs: FS) -> T:
+        raise NotImplementedError
+
+    @classmethod
+    def guess_format(cls, path: str) -> str:
+        _, ext = fs.path.splitext(path)
+
+        ext = ext if ext == "" else ext[1:]
+        if ext in [cls._export_ext(), "gz", "xz", "bz2", "tar", "zip"]:
+            return ext
+        else:
+            return cls._export_ext()
+
+    @classmethod
+    def import_from(
+        cls: t.Type[T],
+        path: str,
+        input_format: t.Optional[str] = None,
+        *,
+        protocol: t.Optional[str] = None,
+        user: t.Optional[str] = None,
+        passwd: t.Optional[str] = None,
+        params: t.Optional[t.Dict[str, str]] = None,
+        subpath: t.Optional[str] = None,
+    ) -> T:
+        try:
+            parsedurl = fs.opener.parse(path)
+        except fs.opener.errors.ParseError:
+            if protocol is None:
+                protocol = "osfs"
+                resource = path if os.sep == "/" else path.replace(os.sep, "/")
+        else:
+            if any(v is not None for v in [protocol, user, passwd, params, subpath]):
+                raise ValueError(
+                    "An FS URL was passed as the output path; all additional information should be passed as part of the URL."
+                )
+
+            protocol = parsedurl.protocol
+            user = parsedurl.username
+            passwd = parsedurl.password
+            resource = parsedurl.resource
+            params = parsedurl.params
+            subpath = parsedurl.path
+
+        if protocol not in fs.opener.registry.protocols:
+            if protocol == "s3":
+                raise ValueError(
+                    "Tried to open an S3 url but the protocol is not registered; did you 'pip install fs-s3fs'?"
+                )
+            else:
+                raise ValueError(
+                    f"Unknown or unsupported protocol {protocol}. Some supported protocols are ftp, s3, and osfs."
```

Comment:
```suggestion
                    f"Unknown or unsupported protocol {protocol}. Some supported protocols are ftp, s3, and OS Filesystem."
```

suggest this change in error message since "osfs" probably don't mean anything to the end user

Reply:
Ah, the protocol is supposed to be `"osfs"`, because that's the name of the FS protocol. Maybe I should add quotes?

Reply:
got it, sounds good!